sst/SimengCoreWrapper.cc

```diff
@@ -0,0 +1,213 @@
+  // Tick the core and memory interfaces until the program has halted
+  if (!core_->hasHalted() || dataMemory_->hasPendingRequests()) {
+    // Tick the core
+    core_->tick();
+
+    // Tick memory
+    instructionMemory_->tick();
+    dataMemory_->tick();
+
+    iterations_++;
+
+    return false;
+  } else {
+    // Protected method from SST::Component used to end SST simulation
+    primaryComponentOKToEndSim();
+    return true;
+  }
+}
+
+std::vector<std::string> SimengCoreWrapper::splitArgs(std::string argString) {
+  int brCount = 0;
+  std::string newArg = "";
+  std::vector<std::string> arguments = {};
+
+  // Using a custom delimiter, split the argString string into individual
+  // arguments and collate in a vector
+  for (int c = 0; c < argString.length(); c++) {
+    // Recored starting delimiter
+    if (argString.at(c) == '[') {
+      brCount++;
+    } else if (argString.at(c) == ']') {
+      // Ensure ending delimiter has a matching start
+      if (brCount <= 0) {
+        std::cerr << "Invalid argument format, mismatch on delimiters"
+                  << std::endl;
+        exit(1);
+      }
+      arguments.push_back(newArg);
+      newArg = "";
+      brCount--;
+    } else if (argString.at(c) == '\\') {
+      // Identified escape character, read next character as an argument
+      // character
+      c++;
+      newArg += argString.at(c);
+    } else {
+      newArg += argString.at(c);
+    }
+  }
+
+  // Ensure all starting delimiters have been matched
+  if (brCount != 0) {
+    std::cerr << "Invalid argument format, unmatched delimiter" << std::endl;
+    exit(1);
+    return {};
+  }
+
+  return arguments;
+}
+
+void SimengCoreWrapper::fabricateSimengCore() {
+  output_.verbose(CALL_INFO, 1, 0, "Setting up SimEng Core\n");
+
+  // Create the instance of the core to be simulated
+  if (configPath_ != "") {
+    coreInstance_ = std::make_unique<simeng::CoreInstance>(
+        configPath_, executablePath_, executableArgs_);
+  } else {
+    coreInstance_ = std::make_unique<simeng::CoreInstance>(executablePath_,
+                                                           executableArgs_);
+  }
+
+  // Set the SST data memory SimEng should use
+  coreInstance_->setL1DataMemory(dataMemory_);
+
+  // Construct core
+  coreInstance_->createCore();
+
+  // Get remaining simulation objects needed to forward simulation
+  core_ = coreInstance_->getCore();
+  instructionMemory_ = coreInstance_->getInstructionMemory();
+
+  // This check ensure that SST has enough memory to store the entire
+  // processImage constructed by SimEng.
+  if (maxAddrMemory_ < coreInstance_->getProcessImageSize()) {
+    output_.verbose(
+        CALL_INFO, 1, 0,
+        "Error: SST backend memory is less than processImage size. Please "
+        "increase the memory allocated to memHierarchy.memBackend and "
+        "ensure it is consistent with \'max_addr_memory\' and "
+        "\'addr_range_end\'. \n");
+    primaryComponentOKToEndSim();
```

Comment:
Do you need to do this every time before exiting? Just wondering if this is the reason why I still have an sst process running after the simulation ends

Reply:
Yes! This method signals SST to stop the simulation.

Reply:
So if somewhere else within SimEng exit is called does OKToEndSim need to be called?

Reply:
No, it doesn't, as that would terminate SST since the simulation runs inside SST.

Reply:
any updates on this?